zilliqa/src/node.rs

```diff
@@ -450,6 +450,25 @@ impl Node {
             .get_account(address)
     }
 
+    /// Return the `nonce` of an account. If the `block_number` is [BlockNumber::Pending], any pending transactions in
+    /// our pool are also added to the nonce.
+    pub fn get_transaction_count(
+        &self,
+        address: Address,
+        block_number: BlockNumber,
+    ) -> Result<u64> {
+        let nonce = self.get_account(address, block_number)?.nonce;
+
```

Comment:
What happens when you've put the transactions into a block, but that block is not yet confirmed? Will get_account(address, block_number) respect the pending block? If not, you'll get the wrong nonce here and thus think there are no consecutive nonces in the mempool, whereas actually (once the pending block is confirmed) there are ..

Reply:
Oof good point, thank you for spotting. `get_account` won't respect the pending block so I believe your bug exists. Just to be clear - When we build a block proposal, we remove transactions from our `TransactionPool`.

Just thinking out loud - What if we keep track of any pending block `Proposal`s in `Consensus` which we've sent out, but haven't been voted on yet. Then any `block_number = Pending` APIs can query from there instead of the tip of the chain?

Tagging @theo-zil and @DrZoltanFazekas for thoughts as well.

Reply:
Maybe an alternative would be not removing transactions from the mempool until they are mined in a block? We could remove them and re-add them when we're building a `Proposal`?

Reply:
Hmm - I already do this only for intershard transactions (see the bottom of the `vote()` definition), so this honestly sounds like a solution. Just remove the redundant re-insertion of `opaque_transactions` inside `vote()`.

Reply:
I think it'll work - certainly the right approach - but myself, I'd probably mark transactions explicitly as having gone into a block rather than relying on the nonce comparison, partly for visibility (which might be a bad idea, since denormalising), and partly in case someone other than this code spotted that they were invalid and ejected them from the mempool in the meantime, which would result in this bug reoccurring as a heisenbug... but that's just me.

Reply:
I'm not sure what you mean by "relying on the nonce comparison"? I don't think it will be simpler to mark transactions in the mempool as having gone into a proposal, since then we need to deal with the case where they go into the proposal, but don't get mined and that flag needs to be reset.

I've implemented the solution discussed above where we retain transactions in the mempool until they are mined.